File: day04/solve.py
import sys
from pathlib import Path

# Add parent directory to path so we can import utils
sys.path.insert(0, str(Path(__file__).parent.parent))

from utils.utils import find_neighbours8, load_grid


def part1(grid: list[str]):
    total = 0

    for row, line in enumerate(grid):
        for col, char in enumerate(line):
            if char != "@":
                continue

            neighbours = find_neighbours8(grid, row, col)
            num_neighbours = sum(1 for r, c in neighbours if grid[r][c] == "@")
            if num_neighbours < 4:
                total += 1

    return total


def _search_and_remove(
    grid: list[str], removed: set[tuple[int, int]], total: int
) -> int:
    round_total = 0
    to_remove: set[tuple[int, int]] = set()

    for row, line in enumerate(grid):
        for col, char in enumerate(line):
            if char != "@" or (row, col) in removed:
                continue

            neighbours = find_neighbours8(grid, row, col)
            num_neighbours = sum(
                1 for r, c in neighbours if grid[r][c] == "@" and (r, c) not in removed
            )

            if num_neighbours < 4:
                round_total += 1
                to_remove.add((row, col))

    removed.update(to_remove)
    if round_total == 0:
        return total
    else:
        return _search_and_remove(grid, removed, total + round_total)


def part2(grid: list[str]):
    total = 0
    removed: set[tuple[int, int]] = set()

    return _search_and_remove(grid, removed, total)


def main():
    grid = load_grid(False)

    result1 = part1(grid)
    print(result1)

    result2 = part2(grid)
    print(result2)


if __name__ == "__main__":
    main()

File: day04/solve.go
package main

import (
	"fmt"
	"log"

	"github.com/bailey4770/aoc2025/utils"
)

type coordT [2]int

func part1(grid []string) int {
	total := 0

	for row, line := range grid {
		for col, char := range line {
			if char != '@' {
				continue
			}

			neighbours := utils.Neighbours8(grid, row, col)
			numNeighbours := 0

			for _, neighbour := range neighbours {
				if grid[neighbour[0]][neighbour[1]] == '@' {
					numNeighbours++
				}
			}

			if numNeighbours < 4 {
				total++
			}
		}
	}

	return total
}

func searchAndRemove(grid []string, removed map[[2]int]struct{}, total int) int {
	roundTotal := 0
	toRemove := make(map[[2]int]struct{})

	for row, line := range grid {
		for col, char := range line {

			coord := coordT{row, col}
			if _, ok := removed[coord]; ok || char != '@' {
				continue
			}

			neighbours := utils.Neighbours8(grid, row, col)
			numNeighbours := 0

			for _, neighbour := range neighbours {
				neighbourCoord := coordT{neighbour[0], neighbour[1]}
				if _, ok := removed[neighbourCoord]; !ok && grid[neighbour[0]][neighbour[1]] == '@' {
					numNeighbours++
				}
			}

			if numNeighbours < 4 {
				roundTotal++
				toRemove[coord] = struct{}{}
			}
		}
	}

	for coord := range toRemove {
		removed[coord] = struct{}{}
	}

	if roundTotal == 0 {
		return total
	} else {
		return searchAndRemove(grid, removed, total+roundTotal)
	}
}

func part2(grid []string) int {
	total := 0
	removed := make(map[[2]int]struct{})

	return searchAndRemove(grid, removed, total)
}

func main() {
	data, err := utils.LoadString(false, "\n")
	if err != nil {
		log.Fatalf("Error: %v", err)
	}

	res1 := part1(data)
	fmt.Printf("Part 1: %v\n", res1)

	res2 := part2(data)
	if err != nil {
		log.Fatalf("Error: %v", err)
	}
	fmt.Printf("Part 2: %v\n", res2)
}
